core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java

```diff
@@ -1055,15 +1055,19 @@ public TrimResult trimFields(
     final ImmutableBitSet.Builder inputFieldsUsed =
         aggregate.getGroupSet().rebuild();
     // 2. agg functions
+    int index = aggregate.getGroupCount();
     for (AggregateCall aggCall : aggregate.getAggCallList()) {
-      inputFieldsUsed.addAll(aggCall.getArgList());
-      if (aggCall.filterArg >= 0) {
-        inputFieldsUsed.set(aggCall.filterArg);
-      }
-      if (aggCall.distinctKeys != null) {
-        inputFieldsUsed.addAll(aggCall.distinctKeys);
+      if (fieldsUsed.get(index)) {
```

Comment:
I'm not familiar here, just one small question: Why do the index start with aggregate's groupCount? In this case, if the top Project contains ENAME, can it be covered?

Reply:
Because all group by keys are necessary in aggregate no matter if they are used in the consumer.  If the top Project contains ENAME, `fieldsUsed.get(indexOfEname)` will be true.

Reply:
Thanks for the clarify

Reply:
I agree with @ILuffZhe that the new code, while correct, is not straightforward to understand, so perhaps a comment will be useful here. I'd suggest maybe updating the comment line 1057 with something like:
`// 2. agg functions: consider only the ones that are needed according to fieldsUsed`